Performance Improvement for Gold Check

Piinks opened the pull request:
- Noticed in the logs that there was some unnecessary work going on (no need to query cirrus if the last known gold status was completed), rearranged the logic to make it leaner. 
- Also updated the messaging, https://github.com/flutter/flutter/pull/51968 is my validator and the comment wasn't as clear as it could be.
- Moves cron job to every 5 minutes instead of 3
- Re-introduces considering what tests are running to reduce gold checks
  - In #651 this was considered liable to break, but I have implemented it differently now. Instead of only checking framework tests, it is just using the presence of a framework test to indicate that golden file tests are executing for this PR
- Reduced comment noise after initial feedback

Piinks commented on app_dart/lib/src/request_handlers/push_gold_status_to_github.dart:
```diff
@@ -59,7 +59,22 @@ class PushGoldStatusToGithub extends ApiRequestHandler<Body> {
+        lastUpdate.head == pr.head.sha) {
+        log.debug('Completed status already reported for this commit.');
+        // We have already seen this commit and it is completed.
+        continue;
```
This is the case for an overwhelming number of PRs.

CaseyHillers commented on app_dart/lib/src/datastore/cocoon_config.dart:
```diff
@@ -89,13 +89,15 @@ class Config {
-      '[Writing a golden file test for `package:flutter`](https://github.com/flutter/flutter/wiki/Writing-a-golden-file-test-for-package:flutter) '
-      'may also provide guidance for this change.\n\n'
+  String get goldenBreakingChangeMessage =>
+      'Changes to golden files are considered breaking, so consult '
```
nit: are considered breaking changes,

Piinks replied:
```suggestion
      'Changes to golden files are considered breaking changes, so consult '
```

Piinks commented:
I have added a few more changes after touching base with @kjlubick about the amount of traffic on the gold endpoint
@dnfield had some concerns earlier (#651) about checking for framework tests, I've reintroduced this in a different way that is more reliable.
I've also moved the cron job to 5 minutes from 3.

PTAL. :)

Piinks commented:
I have to fix the formatting, but oddly dartfmt is different on my laptop. 😖

dnfield commented on app_dart/lib/src/request_handlers/push_gold_status_to_github.dart:
```diff
@@ -90,53 +74,80 @@ class PushGoldStatusToGithub extends ApiRequestHandler<Body> {
+          await queryCirrusGraphQL(pr.head.sha, cirrusClient, log, 'flutter');
+      for (dynamic check in cirrusChecks) {
+        final String status = check['status'];
+        final String taskName = check['name'];
```
This is reminding me that we should turn off `avoid_as` and set `implicit-casts: false`.

I was going to suggest some ways to avoid dynamic typing here but we can save that for a PR that just changes those lints around.

CaseyHillers commented on app_dart/test/request_handlers/push_gold_status_to_github_test.dart:
```diff
@@ -552,63 +645,6 @@ void main() {
       });
-
-      test(
-          'Completed pull request does not skip follow-up prs with early return',
```
To verify, this was removed because this was the performance issue?

Piinks replied:
That was an error. 😅  It's back.

CaseyHillers commented on app_dart/lib/src/request_handlers/push_gold_status_to_github.dart:
```diff
@@ -131,6 +131,7 @@ class PushGoldStatusToGithub extends ApiRequestHandler<Body> {
                 .createStatus(slug, pr.head.sha, statusRequest);
             lastUpdate.status = statusRequest.state;
             lastUpdate.head = pr.head.sha;
+            print('About to update: ${lastUpdate.pr}, ${lastUpdate.updates}');
```
What's the reasoning for switching from using log statements to prints? IIRC AppEngine will still log stdout, but it doesn't assign a log level.

Piinks replied:
Not intentional, just local testing, it ended up getting pushed. I'm considering this WIP until I can find out how I broke a test. :confused:

Piinks commented:
This is all fixed now. Sorry for the noise. :)
